### PR TITLE
Fixed SEP-10 web auth issue

### PR DESCRIPTION
**Pull request by jopmiddelkamp**
## What is fixed
The default value of the now added a grace period to it which can cause failing SEP-10 validations.

## What did I fix
I changed the default value of the now parameter to `DateTimeOffset.Now` and I move the grace logic to the validate time bounds method.

## BREAKING CHANGES:
`WebAuthentication.BuildChallengeTransaction`:
 - Argument `now` rename to `validFrom`
 - Argument `timeout` renamed to `validFor`

`WebAuthentication.ReadChallengeTransaction`:
 - Argument `now` default value change: `DateTimeOffset.Now.AddSeconds(GracePeriod)` -> `DateTimeOffset.Now`

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs [hunk 1]
```diff
@@ -1691,60 +1513,186 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
...
+        var graceValidFor = validFor + TimeSpan.FromSeconds(WebAuthentication.GracePeriod);
+        var graceValidTill = validFrom.Add(graceValidFor);
+
+        // Act & Assert
+        var result = WebAuthentication.ReadChallengeTransaction(
+            transaction,
+            _serverKeypair.AccountId,
+            HomeDomain,
+            WebAuthDomain,
+            // Start 1 second before the max time bounds + grace period
```
Comment (cuongph87) [inline StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1694]: Doesn't the max time bounds already include the grace period?

**Review by cuongph87 — CHANGES_REQUESTED** [review]
There are 2 places in WebAuthentication.cs where "tx" name is used instead of transaction, but I can't add a comment to those lines directly.

**Comment by jopmiddelkamp** [comment]
> There are 2 places in WebAuthentication.cs where "tx" name is used instead of transaction, but I can't add a comment to those lines directly.

Fixed

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs [hunk 2]
```diff
@@ -1691,60 +1513,186 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
...
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBound()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'validFrom'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = WebAuthentication.BuildChallengeTransaction(
```
[the remark this answers is not in the document]
Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1640]: Good idea fixed, also fixed it for the ReadChallengeTransaction

File: StellarDotnetSdk/WebAuthentication.cs [hunk 3]
```diff
@@ -65,21 +79,28 @@ public static class WebAuthentication
...
+        TimeSpan? validFor = null,
+        Network? network = null,
+        string? clientDomain = null,
+        KeyPair? clientSigningKey = null)
     {
-        if (serverKeypair is null) throw new ArgumentNullException(nameof(serverKeypair));
-        if (clientAccountId is null) throw new ArgumentNullException(nameof(clientAccountId));
+        ArgumentNullException.ThrowIfNull(serverKeypair);
+        ArgumentNullException.ThrowIfNull(clientAccountId);
         if (string.IsNullOrEmpty(homeDomain)) throw new ArgumentNullException(nameof(homeDomain));
```
[the remark this answers is not in the document]
Comment (jopmiddelkamp) [reply StellarDotnetSdk/WebAuthentication.cs:104]: Fixes

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs [hunk 4]
```diff
@@ -1691,60 +1513,186 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
...
+                now: graceValidTill.AddSeconds(1));
+        }, "Challenge transaction expired");
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBoundButWithinGracePeriod()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'validFrom'
```
[the remark this answers is not in the document]
Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1671]: Yes, fixed in all methods

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs [hunk 5]
```diff
@@ -246,20 +234,17 @@ public void TestVerifyChallengeTransactionThrowsIfNotSignedByServer()
...
-        var clientKeypair = KeyPair.Random();
-
-        Network.UseTestNetwork();
-
         var now = DateTimeOffset.Now;
 
-        var tx = WebAuthentication.BuildChallengeTransaction(serverKeypair, clientKeypair.AccountId, HomeDomain,
-            WebAuthDomain, now: now);
-        tx.Sign(clientKeypair);
+        var transaction = WebAuthentication.BuildChallengeTransaction(_serverKeypair, _clientKeypair.AccountId,
```
[the remark this answers is not in the document]
Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs:239]: Formatting fixed

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs (hunk 1 above)
Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1694]: This comment should not be there. Removed.

Comment (cuongph87) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs]: Weird, I still see the comment.

Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs]: Ohh no sorry I confused this with another same comment I removed in the `_ReadChallengeTransaction` method.

To address your initial question: no, the grace period is not included in the minimum and maximum time values. In the validateTimeBounds method, the grace period is added to the calculation, but before that, the values are only what the user configured.

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs [hunk 6]
```diff
@@ -1691,60 +1719,193 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
...
+    public void TestReadChallengeTransactionWithOutOfUpperBoundButWithinGracePeriod()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'now'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = _BuildAndSignChallengeTransaction(
+            validFrom,
```
Comment (cuongph87) [inline StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1868]: Not important, but for function calls that are short enough to occupy 1 line, I'd prefer to keep them in 1 line.

**Review by cuongph87 — APPROVED** [review]
Looks good now.

File: StellarDotnetSdk.Tests/WebAuthenticationTest.cs (hunk 6 above)
Comment (jopmiddelkamp) [reply StellarDotnetSdk.Tests/WebAuthenticationTest.cs:1868]: Alright, let's talk about this another time a bit more.

**Pull request merged**